src/leap/bitmask/core/mail_services.py

```diff
@@ -0,0 +1,668 @@
+          watches and processes the Incoming Queue and saves the processed mail
+          into the matching INBOX.
+    """
+
+    name = 'mail'
+
+    # TODO factor out Mail Service to inside mail package.
+
+    subscribed_to_hooks = ('on_new_keymanager_instance',)
+
+    def __init__(self, basedir):
+        self._basedir = basedir
+        self._soledad_sessions = {}
+        self._keymanager_sessions = {}
+        self._sendmail_opts = {}
+        self._imap_tokens = {}
+        self._smtp_tokens = {}
+        self._active_user = None
+        super(StandardMailService, self).__init__()
+        self.initializeChildrenServices()
+
+    def initializeChildrenServices(self):
+        self.addService(IMAPService(self._soledad_sessions))
+        self.addService(IncomingMailService(self))
+        self.addService(SMTPService(
+            self._soledad_sessions, self._keymanager_sessions,
+            self._sendmail_opts))
+
+    def startService(self):
+        log.msg('Starting Mail Service...')
+        super(StandardMailService, self).startService()
+
+    def stopService(self):
+        super(StandardMailService, self).stopService()
+
+    def startInstance(self, userid, soledad, keymanager):
+        username, provider = userid.split('@')
+
+        self._soledad_sessions[userid] = soledad
+        self._keymanager_sessions[userid] = keymanager
+
+        sendmail_opts = _get_sendmail_opts(self._basedir, provider, username)
+        self._sendmail_opts[userid] = sendmail_opts
+
+        incoming = self.getServiceNamed('incoming_mail')
+        incoming.startInstance(userid)
+
+        def registerIMAPToken(token):
+            self._imap_tokens[userid] = token
+            self._active_user = userid
+            return token
+
+        def registerSMTPToken(token):
+            self._smtp_tokens[userid] = token
+            return token
+
+        d = soledad.get_or_create_service_token('imap')
+        d.addCallback(registerIMAPToken)
+        d.addCallback(
+            lambda _: soledad.get_or_create_service_token('smtp'))
+        d.addCallback(registerSMTPToken)
+        return d
+
+    def stopInstance(self):
+        pass
+
+    # hooks
+
+    def hook_on_new_keymanager_instance(self, **kw):
+        # XXX we can specify this as a waterfall, or just AND the two
+        # conditions.
+        userid = kw['userid']
+        soledad = kw['soledad']
+        keymanager = kw['keymanager']
+
+        # TODO --- only start instance if "autostart" is True.
+        self.startInstance(userid, soledad, keymanager)
+
+    # commands
+
+    def do_status(self):
+        return 'mail: %s' % 'running' if self.running else 'disabled'
+
+    def get_imap_token(self):
+        active_user = self._active_user
+        if not active_user:
+            return defer.succeed('NO ACTIVE USER')
+        token = self._imap_tokens.get(active_user)
+        return defer.succeed("IMAP TOKEN (%s): %s" % (active_user, token))
```

Comment:
The dispatcher builds jsons out of that, but still this string might not be so easy to make really computer readable.

At some point we might need to rethink this. Maybe return just `(active_user, token)` and let the dispatcher build a meaningful json with it.